src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,408 @@
+		this.consumerGroup = consumerGroup;
+	}
+
+	/**
+	 * Specify an {@link InboundMessageMapper} to extract message headers embedded
+	 * into the record data.
+	 *
+	 * @param embeddedHeadersMapper the {@link InboundMessageMapper} to use.
+	 * @since 2.0
+	 */
+	public void setEmbeddedHeadersMapper(InboundMessageMapper<byte[]> embeddedHeadersMapper) {
+		this.embeddedHeadersMapper = embeddedHeadersMapper;
+	}
+
+	@Override
+	protected void onInit() {
+		super.onInit();
+
+		String workerId = UUID.randomUUID().toString();
+		RecordProcessorFactory recordProcessorFactory = new RecordProcessorFactory();
+
+		ConfigsBuilder configsBuilder = new ConfigsBuilder(this.stream, this.consumerGroup,
+				this.kinesisClient, this.dynamoDBClient, this.cloudWatchClient, workerId, recordProcessorFactory);
+		configsBuilder.retrievalConfig().initialPositionInStreamExtended(this.streamInitialSequence);
+		configsBuilder.retrievalConfig().listShardsBackoffTimeInMillis(this.consumerBackoff);
+		configsBuilder.coordinatorConfig().parentShardPollIntervalMillis(this.idleBetweenPolls);
+
+		this.scheduler = new Scheduler(configsBuilder.checkpointConfig(),
+			configsBuilder.coordinatorConfig(),
+			configsBuilder.leaseManagementConfig(),
+			configsBuilder.lifecycleConfig(),
+			configsBuilder.metricsConfig(),
+			configsBuilder.processorConfig(),
+			configsBuilder.retrievalConfig());
+	}
+
+	@Override
+	protected void doStart() {
+		super.doStart();
+		this.executor.execute(() -> this.scheduler.run());
+	}
+
+	/**
+	 * Takes no action by default. Subclasses may override this if they need
+	 * lifecycle-managed behavior.
+	 */
+	@Override
+	protected void doStop() {
+
+		super.doStop();
+		this.scheduler.shutdown();
+
+	}
+
+	private AbstractIntegrationMessageBuilder<Object> prepareMessageForRecord(KinesisClientRecord record) {
+		ByteBuffer data = record.data();
+		byte[] dataArray = new byte[data.remaining()];
+		Object payload = dataArray;
+		data.get(dataArray);
+		Message<?> messageToUse = null;
+
+		if (KclMessageDrivenChannelAdapter.this.embeddedHeadersMapper != null) {
+			try {
+				messageToUse = KclMessageDrivenChannelAdapter.this.embeddedHeadersMapper.toMessage((byte[]) payload);
+
+				payload = messageToUse.getPayload();
+			}
+			catch (Exception e) {
+				logger.warn("Could not parse embedded headers. Remain payload untouched.", e);
+			}
+		}
+
+		AbstractIntegrationMessageBuilder<Object> messageBuilder = getMessageBuilderFactory().withPayload(payload)
+				.setHeader(AwsHeaders.RECEIVED_PARTITION_KEY, record.partitionKey())
+				.setHeader(AwsHeaders.RECEIVED_SEQUENCE_NUMBER, record.sequenceNumber());
+
+		if (messageToUse != null) {
+			messageBuilder.copyHeadersIfAbsent(messageToUse.getHeaders());
+		}
+
+		return messageBuilder;
+	}
+
+	private void performSend(AbstractIntegrationMessageBuilder<?> messageBuilder, Object rawRecord) {
+
+		Message<?> messageToSend = messageBuilder.build();
+		try {
+			sendMessage(messageToSend);
```

Comment:
The `setAttributesIfNecessary(rawRecord, messageToSend);` is missed here for downstream error handling.